### PR TITLE
[AGW] mobilityd: dhcp_client: implement dhcp lease

### PR DESCRIPTION
**Pull request by pshelar**
Following patch add support for DHCP lease to support dynamic
ip address allocation via DHCP. DHCP state monitoring is done
in separate monitoring thread in mobilityD.

## Test Plan

`make test` and `make integ_test`

### CODE REVIEW

File: lte/gateway/python/magma/mobilityd/dhcp_client.py [hunk 1]
```diff
@@ -57,6 +58,10 @@ def __init__(self,
         self._dhcp_notify = dhcp_wait
         self._dhcp_interface = iface
         self._msg_xid = 0
+        self._lease_renew_wait_min = lease_renew_wait_min
+        self.monitor_thread = threading.Thread(target=self._monitor_dhcp_state)
```
Comment (ardzoht) [inline lte/gateway/python/magma/mobilityd/dhcp_client.py]: should we move this outside of the DHCPClient? I feel like it's better instantiating this from the IPAllocator and then passing it as a dependency on the constructor?

File: lte/gateway/python/magma/mobilityd/dhcp_client.py [hunk 2]
```diff
@@ -159,13 +163,40 @@ def release_ip_address(self, mac: MacAddress):
...
+                    else:
+                        time_to_renew = dhcp_record.lease_renew_deadline - now
+                        wait_time = min(wait_time, time_to_renew.total_seconds())
+
+            # default in wait is 30 sec
+            wait_time = max(wait_time, self._lease_renew_wait_min)
+            logging.debug("lease renewal check after: %s sec" % wait_time)
+            self.monitor_thread_event.wait(wait_time)
+            if self.monitor_thread_event.is_set():
+                break;
```
Comment (ardzoht) [inline lte/gateway/python/magma/mobilityd/dhcp_client.py]: remove semicolon

File: lte/gateway/python/magma/mobilityd/dhcp_desc.py [hunk 3]
```diff
@@ -35,7 +35,7 @@ def __init__(self, mac: MacAddress, ip: str,
                  state_requested: DHCPState,
                  state: DHCPState = DHCPState.UNKNOWN,
                  subnet: str = None, server_ip=None, router_ip=None,
-                 lease_time=None, xid=None):
+                 lease_time=0, xid=None):
```
Comment (ardzoht) [inline lte/gateway/python/magma/mobilityd/dhcp_desc.py]: add typing here?

File: lte/gateway/python/setup.py [hunk 4]
```diff
@@ -102,6 +102,7 @@
         'eventlet>=0.24',
         'h2>=3.2.0',
         'hpack>=3.0',
+        'freezegun'
```
Comment (ardzoht) [inline lte/gateway/python/setup.py]: for this lets try to declare it on a specific version of the library, we've seen some issues when there's implicit dependencies on other libraries while installing the magma pkg

**Review by ardzoht — CHANGES_REQUESTED** [review]

File: lte/gateway/python/magma/mobilityd/dhcp_client.py (hunk 1 above)
Comment (pshelar) [reply lte/gateway/python/magma/mobilityd/dhcp_client.py]: some of changes would be covered in other RP refactoring mobilityD, we can revisit it, after that.

**Review by ardzoht — APPROVED** [review]

**Pull request closed without merging**